Added auto completion in the EvalOCLDialog

till-a opened the pull request:
Works for 
- collection operations 
- class members 
- use-commands

Note: Autocompletion is currently integrated into the EvalOCLDialog only and not into the shell.
 As a result, only collection operations and class members are fully supported.

Commit:
Added auto completion in the EvalOCLDialog for Collection operations and class members and autocompletion for use-commands (but not integrated into shell)

Commit:
Fixed Nullpointer in autocompletion initialization

Commit:
Fixed occurrence of multiple autocompletion result lists in the EvalOCLDialog

Commit:
Removed data structures in autocompletion class. Now uses MSystemState to get information about the loaded model

h-man2 reviewed (commented):
Great: you implemented many tests 👍

Empty classes could possibly be replaced by interfaces, but this is a minor issue.

Would be great to accept suggestions by pressing `<enter>`.
Reducing the suggested entries while typing would improve teh usabilty, too!

Commit:
Added documentation in documentation/documentation.md

Commit:
Removed unused class SuggestionForClassNames

till-a commented on use-gui/src/main/java/org/tzi/use/autocompletion/SuggestionForClassName.java:
Yes you are right, it indeed can be dropped. 
It went unused in e157e9ae379e8e8ca5fa1ff11e4bbd33389f0ee8 when the internal maps in the AutoCompletion class got removed and got deleted now in 4c661184e4b8ba67700081cf923224ca9c9dd507.

Commit:
Updated AutocompletionResultTypes: Changed empty abstract classes to interfaces

Commit:
Updated display of autocompletion results: Now depending on caret position and prefix. Also fixed possible null pointer in autocompletion class.

till-a commented on use-gui/src/main/java/org/tzi/use/gui/main/EvalOCLDialog.java:
```diff
@@ -315,4 +359,117 @@ private boolean evaluate(String in, boolean evalTree) {
+            });
+
+            autocompletionScrollPane = new JScrollPane(autocompletionResultList);
+            textPane.add(autocompletionScrollPane, 1);
```
Yes good idea. Since 06c5b06713b64fa6c4f0cb5259c74164d636a6ca the auto completion windows depend on the caret position and the prefix already present in the input text area. In my opinion, this resolves the issue effectively.

Commit:
Added dynamically refreshed suggestions
Updated documentation: CTRL+SPACE is no longer needed
Removed some unnecessary comments

till-a commented:
> Would be great to accept suggestions by pressing `<enter>`. 
> Reducing the suggested entries while typing would improve teh usabilty, too!

Both of these should work in 9739f1d11b57d80259d2a1a62cbf5c692b78b6c8 as the suggestions are now dynamically updated with every written character. To move through the suggested list the ArrowDown and ArrowUp keys can be used (also see documentation/documentation.md for keybinds).

Commit:
Refactored autocompetion testsystem
Fixxed nullpointer when there were no suggestions
Changed suggestion behaviour for = statements

ichxorya commented:
May I ask how's the update going, since it's nearly a year from your last commit.
